flink-ml-lib/src/main/java/org/apache/flink/ml/common/util/QuantileSummary.java

```diff
@@ -0,0 +1,400 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.common.util;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.util.Preconditions;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Helper class to compute an approximate quantile summary. This implementation is based on the
+ * algorithm proposed in the paper: "Space-efficient Online Computation of Quantile Summaries" by
+ * Greenwald, Michael and Khanna, Sanjeev. (https://doi.org/10.1145/375663.375670)
+ */
+@Internal
+public class QuantileSummary implements Serializable {
+
+    /** The target relative error. */
+    private final double relativeError;
+
+    /**
+     * The compression threshold. After the internal buffer of statistics crosses this size, it
+     * attempts to compress the statistics together.
+     */
+    private final int compressThreshold;
+
+    /** The count of all the elements inserted to be calculated. */
+    private final long count;
+
+    /** A buffer of quantile statistics. */
+    private final List<StatsTuple> sampled;
+
+    /** The default size of head buffer. */
+    private static final int DEFAULT_HEAD_SIZE = 50000;
+
+    /** The default compression threshold. */
+    private static final int DEFAULT_COMPRESS_THRESHOLD = 10000;
+
+    /** A buffer of the latest samples seen so far. */
+    private List<Double> headBuffer = new ArrayList<>(DEFAULT_HEAD_SIZE);
```

Comment:
nit: Do you think making it a `double[]` could be more memory efficient? I am also fine the keep the current implementation.

Reply:
Indeed `double[]` is more memory efficient, but after repeated thinking I think leave it as an empty ArrayList is better, it is more friendly to small dataset, which is the same to Spark.